File: rosbridge/benchmark_normal.js
const ROSLIB = require('roslib');
const fs = require('fs');

// Connect to ROS 2 via Rosbridge
const ros = new ROSLIB.Ros({
    url: 'ws://localhost:9090'  // Change if Rosbridge runs on another machine
});

// Handle connection events
ros.on('connection', () => console.log('Connected to ROS2 via Rosbridge!'));
ros.on('error', (error) => console.error('Error:', error));
ros.on('close', () => console.log('Connection to ROS2 closed.'));

// Frequency and message settings
const hz = 200;  // 100Hz frequency
const messagesToSend = 100000;  // Number of messages to send

let idNameMapInt = {};
for (let i = 0; i < messagesToSend; i++) {
    idNameMapInt[i.toString()] = i;
}

function getNanoSecTime() {
  var hrTime = process.hrtime();
  return hrTime[0] * 1000000000 + hrTime[1];
}

// Function to generate random trajectory messages
function constructMessages() {
    let messages = [];

    for (let _ in idNameMapInt) {
        let velArray = Array(6).fill().map(() => (Math.random() * 10 - 5));  // Random floats
        messages.push(velArray);
    }
    return messages;
}

// Construct messages
let messages = constructMessages();

// Tracking messages
let expectedIncomingQueue = [];
let invalidIdsNeverArrived = {};
let invalidIdsArrivedTooLate = {};
let arrivedMessages = {};

// Publisher for sending trajectory messages
let velocityPublisher = new ROSLIB.Topic({
    ros: ros,
    name: '/velocity',
    messageType: 'std_msgs/Float64MultiArray'
});

// Function to send messages
function sendMessages() {
    // Send messages at regular intervals
    let messageIndex = 0;
    
    function sendNextMessage() {
        if (messageIndex >= messagesToSend) {
            console.log("All messages sent!");
            return;
        }
        
        const idNum = messageIndex;
        const sentTime = getNanoSecTime();
        
        expectedIncomingQueue.push({ idNum, sentTime });

        let message = new ROSLIB.Message({
            data: messages[messageIndex]
        });

        velocityPublisher.publish(message);
        
        messageIndex++;
        setTimeout(sendNextMessage, 1000 / hz);
    }
    
    // Start sending
    sendNextMessage();
}
// Subscriber to listen for `JointState` messages
let jointStateSubscriber = new ROSLIB.Topic({
    ros: ros,
    name: '/joint_states',
    messageType: 'sensor_msgs/JointState'
});

// Handling received messages
jointStateSubscriber.subscribe((msg) => {
    let currentTime = getNanoSecTime();
    let { idNum, sentTime } = expectedIncomingQueue.shift();
    arrivedMessages[idNum] = { sentTime, ros2Time: msg.header.stamp, receivedTime: currentTime };
});

// Start sending messages
ros.on('connection', async() => {
    sendMessages();
    await new Promise(resolve => setTimeout(resolve, 600000));
    jointStateSubscriber.unsubscribe();
    saveToCSV();
});

// Function to save CSV data
function saveToCSV() {
    let arrivedCSV = "ID,Sent Time,ROS2 Time,Received Time\n";
    for (let id in arrivedMessages) {
        let { sentTime, ros2Time, receivedTime } = arrivedMessages[id];
        let ns = ros2Time.nanosec.toString().padStart(9, '0');
        arrivedCSV += `${id},${sentTime},${ros2Time.sec}${ns},${receivedTime}\n`;
    }

    let neverArrivedCSV = "ID,Sent Time\n";
    for (let id in invalidIdsNeverArrived) {
        neverArrivedCSV += `${id},${invalidIdsNeverArrived[id]}\n`;
    }

    let tooLateCSV = "ID,Sent Time,ROS2 Time,Received Time\n";
    for (let id in invalidIdsArrivedTooLate) {
        let { sentTime, ros2Time, receivedTime } = invalidIdsArrivedTooLate[id];
        tooLateCSV += `${id},${sentTime},${ros2Time.sec}.${ros2Time.nanosec},${receivedTime}\n`;
    }

    fs.writeFileSync("arrived_messages.csv", arrivedCSV);
    fs.writeFileSync("invalid_ids_never_arrived.csv", neverArrivedCSV);
    fs.writeFileSync("invalid_ids_arrived_too_late.csv", tooLateCSV);
}

File: ros2_api/benchmark_normal.py
import time
import queue
import random
import csv
from ros2_sdk.ros2_sdk import ROS2SDK, TrajPoint

hz = 200  # Frequency of messages in Hz
messages_to_send = 100000  # Number of messages to send
sdk = ROS2SDK()
sdk.connect("TCP", {"port_send": 5555, "port_recv": 5556, "ip": "localhost"})
sdk.send_effort([0, 0, 0, 0, 0, 0], "test")

id_name_map_int = {f"{i}": i for i in range(messages_to_send)}  

def construct_messages(): 
    messages = []
    for id_str, _ in id_name_map_int.items():
        vel_array = [random.uniform(-5, 5) for _ in range(6)]
        messages.append(vel_array)
    return messages
        
messages = construct_messages()

expected_incoming_queue = queue.Queue()
invalid_ids_never_arrived = dict()  
invalid_ids_arrived_too_late = dict()
arrived_messages = dict()

def send_messages():
    for _, id_num in id_name_map_int.items():
        expected_incoming_queue.put((id_num, time.time_ns())) 
        sdk.send_velocity(messages[id_num], name="velocity") 
        time.sleep(1 / hz)  # Sleep for the desired frequency



def on_message_arrival(msg):
    current_time = time.time_ns()  
    id_num, sent_time = expected_incoming_queue.get()
    arrived_messages[id_num] = (sent_time, current_time)
    return
            

subscription = sdk.get_state_stream().subscribe(on_message_arrival)
send_messages()

time.sleep(5) # Wait for messages to arrive
sdk.write_time_data_to_csv()
subscription.dispose()

def write_csv_arrived(filename, data, headers):
    """ Writes dictionary data to a CSV file. """
    with open(filename, 'w', newline='') as csvfile:
        writer = csv.writer(csvfile)
        writer.writerow(headers)
        for id, (sent_time, received_time) in data.items():
            writer.writerow([id, sent_time, received_time])
            
def write_csv_missing(filename, data, headers):
    """ Writes dictionary data to a CSV file. """
    with open(filename, 'w', newline='') as csvfile:
        writer = csv.writer(csvfile)
        writer.writerow(headers)
        for id, (sent_time) in data.items():
            writer.writerow([id, sent_time])

        
write_csv_arrived("arrived_messages.csv", arrived_messages, ["ID", "Sent Time", "Received Time"])
write_csv_missing("invalid_ids_never_arrived.csv", invalid_ids_never_arrived, ["ID", "Sent Time"])  
write_csv_arrived("invalid_ids_arrived_too_late.csv", invalid_ids_arrived_too_late, ["ID", "Sent Time", "Received Time"])